app/src/editor/canvas/components/CodeEditor/index.jsx

```diff
@@ -1,101 +1,114 @@
-            editorOpen: false,
-        })
-    }
-
-    onShowDebug = () => {
-        this.setState({
-            debugOpen: true,
-        })
-    }
-
-    onCloseDebug = () => {
-        this.setState({
-            debugOpen: false,
-        })
-    }
-
-    onEditorPositionUpdate = (x, y) => {
-        this.setState({
-            editorPosition: {
-                x,
-                y,
-            },
-        })
-    }
-
-    onDebugPositionUpdate = (x, y) => {
-        this.setState({
-            debugPosition: {
-                x,
-                y,
-            },
-        })
-    }
-
-    render() {
-        const {
-            readOnly,
-            code,
-            debugMessages,
-            onApply,
-            onChange,
-            onClearDebug,
-            children,
-        } = this.props
-        const { editorOpen, debugOpen, editorPosition, debugPosition } = this.state
-
-        return (
-            <React.Fragment>
-                {children && children(this.onShowEditor)}
-                {!!editorOpen && (
-                    <CodeEditorWindow
-                        position={editorPosition}
-                        onPositionUpdate={this.onEditorPositionUpdate}
-                        code={code}
-                        readOnly={readOnly}
-                        onClose={this.onCloseEditor}
-                        onApply={onApply}
-                        onChange={onChange}
-                        onShowDebug={this.onShowDebug}
-                    />
-                )}
-                {!!debugOpen && (
-                    <DebugWindow
-                        onPositionUpdate={this.onDebugPositionUpdate}
-                        position={debugPosition}
-                        messages={debugMessages}
-                        onClear={onClearDebug}
-                        onClose={this.onCloseDebug}
-                    />
-                )}
-            </React.Fragment>
-        )
-    }
+    return (
+        <React.Fragment>
+            {children && children(onShowEditor)}
+            {!!editorOpen && (
+                <CodeEditorWindow
+                    position={{
+                        x: editorLayout.x,
+                        y: editorLayout.y,
+                    }}
+                    onPositionUpdate={onEditorPositionUpdate}
+                    size={{
+                        width: editorLayout.width,
+                        height: editorLayout.height,
+                    }}
+                    onSizeUpdate={onEditorSizeUpdate}
+                    code={code}
+                    readOnly={readOnly}
+                    onClose={onCloseEditor}
+                    onApply={onApply}
+                    onChange={onChange}
+                    onShowDebug={onShowDebug}
+                />
+            )}
+            {!!debugOpen && (
+                <DebugWindow
+                    onPositionUpdate={onDebugPositionUpdate}
+                    position={{
```

Comment:
possible could just pass `debugLayout` for both `position` and `size`, would pass strict equality test on render